### PR TITLE
Implement frameless window

### PR DESCRIPTION
**Pull request by f-squirrel**
Changes:
* Add function `GuiWindowFrameless`
* Update neovim with `g:GuiWindowFrameless`
* Update help
* Update tests

Fixes https://github.com/equalsraf/neovim-qt/issues/954.

@jgehrig , please review :)

### CODE REVIEW

File: src/gui/shell.cpp
Comment (jgehrig) [inline src/gui/shell.cpp:843]: nitpick: Use the real type + a descriptive name.

```suggestion
				const bool isWindowFrameOn{ variant_not_zero(args.at(1)) };
```

File: src/gui/shell.cpp
Comment (jgehrig) [inline src/gui/shell.cpp:841]: Please move this block into a small function such as `handleWindowFrameless`.

You should be able to follow a similar pattern to `handleLineSpace` below.

File: src/gui/mainwindow.cpp
Comment (jgehrig) [inline src/gui/mainwindow.cpp:199]: Let's avoid the `foo ? 1 :0` ternary (if it is not too much effort).

You might be able to leverage `variant_not_zero(set)`

Comment (jgehrig) [reply src/gui/mainwindow.cpp:199]: Here and elsewhere

File: src/gui/mainwindow.cpp
Comment (jgehrig) [inline src/gui/mainwindow.cpp:192]:
```suggestion
void MainWindow::neovimFrameless(bool isFrameless)
```

nitpick: The name `set` is a little ambiguous.

**Comment by jgehrig** [comment]
@f-squirrel 

This is awesome! Thank you for putting this Pull Request together :clap:

I've added some small insignificant comments. The code looks good. I will do some testing and then merge the changes. We should test on all platforms: Linux, Windows, and MacOS.

File: src/gui/mainwindow.cpp
Comment (f-squirrel) [reply src/gui/mainwindow.cpp]: done

File: src/gui/shell.cpp
Comment (f-squirrel) [reply src/gui/shell.cpp:841]: done

File: src/gui/mainwindow.cpp
Comment (f-squirrel) [reply src/gui/mainwindow.cpp]: `vim_set_var(str, bool)` generates `v:bool` in vimscript while the variable and function accept integers (I did it the same way as `GuiWindowMazimized` and others).

**Comment by f-squirrel** [comment]
> @f-squirrel
> 
> This is awesome! Thank you for putting this Pull Request together clap
> 
> I've added some small insignificant comments. The code looks good. I will do some testing and then merge the changes. We should test on all platforms: Linux, Windows, and macOS.

Thank for the review, please see my answers and fixes in the new commit.
The feature has been tested on Ubuntu 20.04; unfortunately, I don't have macOS or Windows.

P.S. Clazy Github action failed on downloading some dependency.

**Comment by f-squirrel** [comment]
hey folks, @jgehrig and @equalsraf , would you mind reviewing my changes?

**Comment by jgehrig** [comment]
@f-squirrel

Your changes have my approval! :+1: 

The final hold-up is testing on MacOS/Windows. I will merge once someone has tested on the other platforms.

It is on my TODO list, but life is getting the better of me right now :)

File: src/gui/mainwindow.cpp
Comment (jgehrig) [inline src/gui/mainwindow.cpp:195]: Pure curiosity... What does "DD:" stand for?

I've never run across this acronym  before. Totally fine to keep as-is!

**Review by jgehrig — APPROVED** [review]

File: src/gui/mainwindow.cpp
Comment (f-squirrel) [reply src/gui/mainwindow.cpp:195]: DD is my initials (Dmitry Danilov), sorry for the confusion.

**Comment by f-squirrel** [comment]
@jgehrig 

I have tested the PR on macOS with a little help from my friend @little-zaets.

Performed tests:
1. The initial value of `g:GuiWindowFrameless` is `0`
1. Call `GuiWindowFrameless(1)` -> The window becomes frameless
1. The value of `Gg:uiWindowFrameless` is `1`
1. Call `GuiWindowFrameless(0)` -> The window frame is shown
1. The value of `g:GuiWindowFrameless` is `0`

Please find the screen recording below:
nvim

macOS info:
photo_2022-02-11_19-35-52

P.S. Windows is more problematic because I haven't worked with it for ages.

**Comment by jgehrig** [comment]
@f-squirrel

Sorry for the delay.

I quickly tested the feature on MacOS and Windows. Both look good.

However, I do see one issue on Linux:
Linux-Issue-Cursor-Focus

I'm running KDE/Plasma on the machine displaying the focus issues. The focus issue goes away after cycling the focus.

1. Do you see these issues?
2. What Desktop Environment are you using?

We should investigate and fix if possible.

I don't think this issue is a check-in blocker... If you cannot reproduce the issue or we cannot find an easy solution, we can file an Issue to track this; the problem is minor.

**Comment by f-squirrel** [comment]
@jgehrig ,
No problem!

I do not see these issues, to be honest.
Maybe the application loses focus once we change the style?
Not sure how to debug it.

I am using Ubuntu 20.04, please find more info below:
screen-fetch

**Comment by jgehrig** [comment]
> I do not see these issues, to be honest.

No worries. This is actually what I had hoped to hear!

I'm guessing this issue is specific to my Desktop Environment (Plasma/KDE), since you're running GNOME.

>  Maybe the application loses focus once we change the style?

On my machine, the window has focus but calls to `hasFocus()` inside `ShellWidget` return `false`.

It looks like we receive a `QFocusEvent(FocusOut, ActiveWindowFocusReason)` when the `Qt::FramelessWindowHint` flag is set. The Window regains focus on `show()`, but there is no corresponding `FocusIn` event for ShellWidget.

I tried quickly calling some functions to explicitly force focus into `QMainWindow` and `ShellWidget`, but this has no effect. This smells like a Qt or Plasma bug to me.

> Would you mind trying to reproduce it with the following:
>
> NVIM_QT_RUNTIME_PATH=/usr/local/share/nvim-qt/runtime nvim-qt -- -i NONE -u NONE

The issue still reproduces.

File: src/gui/mainwindow.cpp
Comment (jgehrig) [reply src/gui/mainwindow.cpp:199]: A quick update to this block...

- `setWindowFlag` is not supported in the minimum Qt version, 5.8.
- A reference to the focus issue I observed.
- More details on how `setWindowFlags(...)` and `show()` interact.

**Comment by f-squirrel** [comment]
hey @jgehrig,
I have applied the changes you have proposed!

**Comment by khalidchawtany** [comment]
On Mac OS, the frameless window cannot be dragged nor resized. It cannot be resized in any way, not using the cursor, nor using a window manager.

**Comment by khalidchawtany** [comment]
Now reading my comment I sound like a d..k. Sorry about that.

**Comment by jgehrig** [comment]
@khalidchawtany

No worries! (I did not think your comment was obnoxious)

Agreed... Your patch is functionally better. However, it has cross-platform maintainability drawbacks: How do we implement equivalent behavior for Windows/Linux?

1. It is awesome to have a link to your commit here! I am certain it will be useful to MacOS users in the future. :+1: 
2. The commit is not check-in ready for general audiences
  a. No Windows/Linux equivalent implementation yet.
  b. I prefer to avoid adding new languages and platform specific code whenever possible.

**Comment by jgehrig** [comment]
@f-squirrel 

Thanks for the ping... I'm just slow :)

**Pull request merged**
